### DISCUSSION TITLE
DB Lock의 종류

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#178] [Database]
.

**Comment by Kimprodp** [comment]
**[DB Lock]**
데이터베이스는 여러 사용자나 프로세스가 같은 데이터를 동시에 접근하는 상황에서 데이터의 무결성과 일관성을 지키기 위해 락을 사용한다.

**[DB Lock의 종류]**
- 공유 락 (Shared Lock) : 데이터를 변경하지 않는 읽기 명령에 대해 주어지는 락으로 여러 트랜잭션이 동일한 데이터를 읽을 수 있도록 허용한다. (공유 락 끼리는 동시에 접근이 가능하다.)
- 베타 락 (Exclusive Lock) : 데이터에 변경을 가하는 쓰기 명령들에 주어지는 락으로 다른 트랜잭션이 데이터에 접근하는 것을 막는다. 베타 락이 설정된 데이터는 다른 트랜잭션이 읽거나 수정할 수 없으며, 베타 락은 트랜잭션 동안 유지된다.
- 업데이트 락 (Update Lock) : 데이터를 수정하기 위해 베타 락을 걸기 전, 데드 락을 방지하기 위해 사용되는 락이다. 일반적으로 업데이트 락은 UPDATE 쿼리의 필터(WHERE)가 실행되는 과정에서 적용된다. 데이터 수정 전, 다른 트랜잭션이 해당 데이터에 공유 락을 설정하는 것을 방지한다.
- 내재 락 (Intent Lock) : 상위 레벨의 데이터 잠금을 나타내기 위해 사용되며, 하위 레벨에서 잠금을 설정하기 전에 상위 레벨에도 잠금을 설정하기 위해 사용한다.

**Comment by seunghye218** [comment]
## DBMS Lock 종류

1. Exclusive (X) Lock:
    - 가장 강력한 락 유형
    - 데이터 수정(쓰기, 삭제, 갱신)을 위해 사용
    - 한 번에 하나의 트랜잭션만 X Lock을 획득 가능
    - 다른 모든 락 유형과 호환되지 않음
    - 예: UPDATE, DELETE 작업 시 사용
2. Shared (S) Lock:
    - 데이터 읽기 작업에 사용
    - 여러 트랜잭션이 동시에 S Lock을 획득 가능
    - X Lock과는 호환되지 않음
    - 예: SELECT 문 실행 시 사용
3. Intent exclusive (IX) Lock:
    - 하위 레벨에서 X Lock을 획득하려는 의도를 나타냄
    - 주로 테이블이나 페이지 레벨에서 사용
    - S Lock과 호환되지 않음
    - 다른 IX Lock과는 호환됨
    - 예: 테이블의 특정 행에 X Lock을 걸기 전에 테이블에 IX Lock 설정
4. Intent shared (IS) Lock:
    - 하위 레벨에서 S Lock을 획득하려는 의도를 나타냄
    - 주로 테이블이나 페이지 레벨에서 사용
    - X Lock과 호환되지 않지만, 다른 모든 Intent Lock과 호환됨
    - 예: 테이블의 특정 행에 S Lock을 걸기 전에 테이블에 IS Lock 설정
5. Shared intent exclusive (SIX) Lock:
    - S Lock과 IX Lock의 조합
    - 전체 리소스에 대한 S Lock과 일부 하위 리소스에 대한 X Lock 의도를 나타냄
    - S Lock과 호환되지만, 다른 모든 락 유형과 호환되지 않음
    - 예: 테이블의 모든 데이터를 읽으면서 일부 행을 수정하는 경우

Lock 유형들의 주요 목적:

1. 세분화된 동시성 제어
2. 락 에스컬레이션 방지
3. 데이터 일관성 유지
4. 트랜잭션 격리 수준 구현

개발자 관점에서 이해해야 할 점:

- 각 락 유형의 특성과 호환성을 이해하고 적절히 사용해야 함
- 락 경합(Lock contention)을 최소화하여 성능 저하 방지
- 데드락(Deadlock) 가능성을 고려한 설계 필요
- 트랜잭션 설계 시 필요한 최소한의 락만 사용하도록 주의

## Lock 단점

1. 복구 가능성에서 자유롭지 않을 수 있음
2. 교착 상태에서 자유롭지 않음
3. 기아 상태에서 자유롭지 않음
4. cascading 롤백에서 자유롭지 않음
